c/driver/postgresql/connection.cc

```diff
@@ -35,19 +36,116 @@ static const uint32_t kSupportedInfoCodes[] = {
     ADBC_INFO_DRIVER_VERSION, ADBC_INFO_DRIVER_ARROW_VERSION,
 };
 
+struct PqRecord {
+  const char* data;
+  const int len;
+  const bool is_null;
+};
+
+class PqResultRow {
+ public:
+  PqResultRow(pg_result* result, int row_num) : result_(result), row_num_(row_num) {
+    ncols_ = PQnfields(result);
+  }
+  class iterator {
+    const PqResultRow& outer_;
+    int col_num_ = 0;
+
+   public:
+    explicit iterator(const PqResultRow& outer, int col_num = 0)
+        : outer_(outer), col_num_(col_num) {}
+    iterator& operator++() {
+      col_num_++;
+      return *this;
+    }
+
+    iterator operator++(int) {
+      iterator retval = *this;
+      ++(*this);
+      return retval;
+    }
+
+    iterator operator+=(const int& n) {
+      col_num_ += n;
+      return *this;
+    }
+
+    iterator operator+(const int& n) {
+      iterator tmp(*this);
+      tmp += n;
+      return tmp;
+    }
+
+    PqRecord operator[](const int& n) { return *(*this + n); }
+
+    bool operator==(iterator other) const { return col_num_ == other.col_num_; }
```

Comment:
Not sure if C++ would require an equality comparison of the containing class as well. Excluded for now to cut down on verbosity but someone with better C++ knowledge likely knows more